alphadia/fdrexperimental.py

```diff
@@ -107,6 +227,383 @@ def from_state_dict(self, state_dict: dict):
+        group_columns: list[str] | None = None,
+    ) -> pd.DataFrame:
+        """
+        Return dataframe including only the found precursors.
+        """
+        df = df_.copy()
+        df = apply_absolute_transformations(df)
+
+        if self.first_classifier.fitted:
+            df["proba"] = self.first_classifier.predict_proba(df[x_cols].to_numpy())[
+                :, 1
+            ]
+            df_subset = get_entries_below_fdr(
+                df, self.first_fdr, group_columns, remove_decoys=False
+            )
+            print(
+                f"After q-val extraction, after LinClassifier (first_fdr={self.first_fdr}): {df_subset.shape}"
+            )
+
+            self.second_classifier.batch_size = 500
+            self.second_classifier.epochs = 50
+
+            self.second_classifier.fit(
+                df_subset[x_cols].to_numpy(), df_subset[y_col].to_numpy()
+            )
+            df_subset["proba"] = self.second_classifier.predict_proba(
+                df_subset[x_cols].to_numpy()
+            )[:, 1]
+
+        else:
+            df_train = df[df["rank"] < self.train_on_top_n]
+            self.second_classifier.fit(
+                df_train[x_cols].to_numpy(),
+                df_train[y_col].to_numpy(),
+            )
+
+            df_subset = df
+            x_subset = df_subset[x_cols].to_numpy()
+            df_subset["proba"] = self.second_classifier.predict_proba(x_subset)[:, 1]
+
+        df_subset = get_entries_below_fdr(
+            df_subset, self.second_fdr, group_columns
+        )  # , remove_decoys=True)
+        print(
+            f"After q-val extraction, after NN (second_fdr={self.second_fdr}): {df_subset.shape}"
+        )
+
+        df_subset_2 = get_target_decoy_partners(df_subset, df_)
+        df_targets = df_subset_2[df_subset_2["decoy"] == 0]
+
+        self._update_classifier(
+            self.first_classifier,
+            df_subset_2,
+            x_cols,
+            y_col,
+            self.first_fdr,
+            group_columns,
+        )
+
+        return df_targets
+
+    @classmethod
+    def _update_classifier(
+        cls, classifier, df_, x_cols, y_col, fdr, group_columns
+    ) -> None:
+        X = df_[x_cols]
+        y = df_[y_col]
+        df = df_.copy()
+        if hasattr(classifier, "fitted") and classifier.fitted:
+            df["proba"] = classifier.predict_proba(df[x_cols].to_numpy())[:, 1]
+            df_targets = get_entries_below_fdr(df, fdr, group_columns)
+            previous_n_precursors = len(df_targets)
+            saved_state_dict = classifier.to_state_dict()
+        else:
+            previous_n_precursors = -1
+
+        classifier.fit(X, y)
+        classifier._fitted = True
+
+        df["proba"] = classifier.predict_proba(df[x_cols].to_numpy())[:, 1]
+        df_targets = get_entries_below_fdr(df, fdr, group_columns)
+
+        current_n_precursors = len(df_targets)
+        if previous_n_precursors > current_n_precursors:
```

Comment:
At the moment we have to run the first classifier twice. Could we just store the value from the first run and reuse it here?